runtime/type.js

```diff
@@ -10,12 +10,78 @@
 const assert = require('assert');
 const typeLiteral = require('./type-literal.js');
 
+let nextVariableId = 0;
 class Type {
   constructor(key) {
+    assert(typeof key != 'string')
     assert(!typeLiteral.isNamedVariable(key));
     this.key = key;
   }
+
+  // Replaces 'prevariable' types with 'variable'+id types .
+  static assignVariableIds(literal, variableMap) {
```

Comment:
Add a TODO that this should be an operation on types rather than on literals (and ditto for schema fixup below)